JSTL-InstructorCourseEditPage

bamboo3250 opened the pull request:
@yj-soh please review this PR. The original JSP was quite long so I had to split it into a few tag files.

ps:  got dev-green

yj-soh commented on src/main/java/teammates/ui/controller/InstructorCourseEditPageAction.java:
```diff
@@ -32,26 +32,26 @@ public ActionResult execute() throws EntityDoesNotExistException {
-        data.course = courseToEdit;
+        
+        List<InstructorAttributes> instructorList = new ArrayList<InstructorAttributes>();
+        int offset = -1;
```
Should this be changed to `offset`?
Is it offsetting something or is it simply an index? (is it added on / subtracted from something somewhere?)
If `index` is more fitting, maybe change the `index` above to `indexStr` or something.

Also, a comment for why it is `-1` would be good. (e.g. "-1 for new instructor.", etc)

bamboo3250 replied:
offset is not really an index. It's more like a starting point. If it is -1 then show all instructors, otherwise, show the instructor at the position __offset__.

yj-soh commented on src/main/java/teammates/ui/controller/InstructorCourseEditPageData.java:
```diff
@@ -7,23 +8,119 @@
+    public List<CourseEditInstructorPanel> instructorPanelList;
+    public CourseEditInstructorPanel addInstructorPanel;
+    public InstructorAttributes currentInstructor;
+    public ElementTag addInstructorButton;
```
Should be `private`?

bamboo3250 replied:
You mean the class or its attributes?

yj-soh replied:
The attributes.

yj-soh commented on src/main/java/teammates/ui/controller/InstructorCourseEditPageData.java:
```diff
@@ -7,23 +8,119 @@
+        boolean isDisabled = !currentInstructor.isAllowedForPrivilege(Const.ParamsNames.INSTRUCTOR_PERMISSION_MODIFY_COURSE);
+        deleteCourseButton = createButton("<span class=\"glyphicon glyphicon-trash\"></span>Delete", "btn btn-primary btn-xs pull-right",
+                                          "courseDeleteLink", getInstructorCourseDeleteLink(this.course.id, false), Const.Tooltips.COURSE_DELETE,
+                                          "return toggleDeleteCourseConfirmation('" + this.course.id + "');", isDisabled);
```
`course` is equivalent to `this.course` by this point; there's no need for `this` when it is unnecessary.

yj-soh commented on src/main/java/teammates/ui/controller/InstructorCourseEditPageData.java:
```diff
@@ -7,23 +8,119 @@
+                                          "return toggleDeleteCourseConfirmation('" + this.course.id + "');", isDisabled);
+        
+        isDisabled = !currentInstructor.isAllowedForPrivilege(Const.ParamsNames.INSTRUCTOR_PERMISSION_MODIFY_INSTRUCTOR);
+        addInstructorButton = createButton(null, "btn btn-primary", "btnShowNewInstructorForm", null, null, "showNewInstructorForm()", isDisabled);
```
Keep character limit per line in mind, please.

yj-soh commented on src/main/java/teammates/ui/controller/InstructorCourseEditPageData.java:
```diff
@@ -7,23 +8,119 @@
+                                                      null, isDisabled);
+            
+            buttonContent = "<span class=\"glyphicon glyphicon-trash\"></span> Delete";
+            isDisabled = !this.currentInstructor.isAllowedForPrivilege(Const.ParamsNames.INSTRUCTOR_PERMISSION_MODIFY_INSTRUCTOR);
```
`isDisabled` shouldn't change at all between these buttons, should it?

yj-soh commented on src/main/java/teammates/ui/controller/InstructorCourseEditPageData.java:
```diff
@@ -7,23 +8,119 @@
+            instructorList.add(instructorPanel.instructor);
+        }
+        return instructorList;
+    }
```
Is this method used anywhere except in the tests?
If the tests require the data in a certain format, it should convert the data itself (i.e. call `getInstructorPanelList()` and compare each of the `instructor` strings from them).

yj-soh commented on src/main/java/teammates/ui/controller/InstructorCourseEditPageData.java:
```diff
@@ -7,23 +8,119 @@
-
     public String getInstructorCourseInstructorDeleteLink(String courseId, String instructorEmail) {
         String link = Const.ActionURIs.INSTRUCTOR_COURSE_INSTRUCTOR_DELETE;
         link = Url.addParamToUrl(link,Const.ParamsNames.COURSE_ID, courseId);
```
Should these two methods be `public`?

yj-soh commented on src/main/java/teammates/ui/template/CourseEditFeedbackSessionRow.java:
```diff
@@ -0,0 +1,22 @@
+
+public class CourseEditFeedbackSessionRow {
+    public String feedbackSessionName;
+    public List<ElementTag> permissionCheckBoxes;
```
`private`?
Likewise for the other `template` classes.

yj-soh commented on src/main/java/teammates/ui/template/CourseEditInstructorPanel.java:
```diff
@@ -0,0 +1,189 @@
+    public InstructorAttributes instructor;
+    public List<ElementTag> permissionInputGroup1;
+    public List<ElementTag> permissionInputGroup2;
+    public List<ElementTag> permissionInputGroup3;
```
Are there more informative names for these?

bamboo3250 replied:
I don't know if I can find meaningful names for them. It looks like they are separated just for displaying purpose only :/

yj-soh commented on src/main/java/teammates/ui/template/CourseEditSectionRow.java:
```diff
@@ -0,0 +1,187 @@
+    public String sectionName;
+    public InstructorAttributes instructor;
+    public List<ElementTag> permissionInputGroup2;
+    public List<ElementTag> permissionInputGroup3;
```
Now there's only 2 and 3?

bamboo3250 replied:
yeah, group 2 & 3 here are equivalent to the group 2 & 3 of CourseEditInstructorPanel.java. It's weird to me but this is exactly how it was in the original jsp.

yj-soh commented on src/main/webapp/WEB-INF/tags/instructor/course/courseEditAddInstructorPanel.tag:
```diff
@@ -0,0 +1,285 @@
+                                </div>
+                                
+                                <c:if test="${not empty addInstructorPanel.sectionRows}">
+                                    <c:forEach items="${addInstructorPanel.sectionRows}" var="sectionRow" varStatus="i">
```
Maybe put the `<c:if>` after the `<c:forEach>` like for the `feedbackSessions` below?
This will lower the indentation of this whole block while achieving the same effect.

yj-soh commented on src/main/webapp/WEB-INF/tags/instructor/course/courseEditCourseInfo.tag:
```diff
@@ -0,0 +1,56 @@
+<%@ tag import="teammates.common.util.FieldValidator" %>
+<%@ attribute name="deleteCourseButton" type="teammates.ui.template.ElementTag" required="true" %>
+<%@ attribute name="course" type="teammates.common.datatransfer.CourseAttributes" required="true" %>
+<%@ attribute name="googleId" required="true" %>
```
Is this equivalent to `${data.account.googleId}`?
It should be okay to use it like a constant if so.

bamboo3250 replied:
You mean just use "${data.account.googleId}" instead of making it an attribute of a tag?

yj-soh replied:
Yes.

yj-soh commented on src/test/java/teammates/test/cases/ui/pagedata/InstructorCourseEditPageDataTest.java:
```diff
@@ -0,0 +1,78 @@
+        assertNotNull(pageData.addInstructorButton);
+        
+        assertNotNull(pageData.instructorPanelList);
+        assertEquals(3, pageData.instructorPanelList.size());
```
It may be better to use dynamic values like `instructorList.size()` as the expected output.
This implies that the `instructorList` passed in is correctly interpreted into a `instructorPanelList` (length-wise).

bamboo3250 commented:
@yj-soh ready to get reviewed again.

bamboo3250 commented:
I'll try refactoring this PR one more time.

bamboo3250 commented:
@yj-soh ready for review. Somehow I still want to keep group1, group2 and group3 checkboxes. If you cannot accept that then I'll separate them into 11 checkboxes to show them individually.

yj-soh commented on src/main/java/teammates/ui/template/CourseEditSectionRow.java:
```diff
@@ -0,0 +1,225 @@
+                   + Const.ParamsNames.INSTRUCTOR_SECTION_GROUP + sectionIndex + "feedback" + feedbackName;
+            isChecked = (instructor != null) 
+                        && instructor.isAllowedForPrivilege(sectionName, feedbackName, 
+                                        Const.ParamsNames.INSTRUCTOR_PERMISSION_VIEW_SESSION_IN_SECTIONS);
```
Not sure if it's a good idea, but would it be tidier to have a `Map<String, Boolean>` with all the privileges set beforehand?
You could just list all the constants in an array or something, then loop trough them to populate the map, instead of having 10 of these
` (instructor != null) && instructor.isAllowedForPrivilege(sectionName, feedbackName, Const.ParamsNames.*);`

yj-soh commented on src/main/java/teammates/ui/template/CourseEditSectionRow.java:
```diff
@@ -0,0 +1,225 @@
+            
+            feedbackSessionRow.getPermissionCheckBoxes().add(submitSessionInSectionCheckBox);
+            feedbackSessionRow.getPermissionCheckBoxes().add(viewSessionInSectionCheckBox);
+            feedbackSessionRow.getPermissionCheckBoxes().add(modifySessionInSectionCheckBox);
```
Like earlier, possible to make a method-scoped list and toss it into the constructor?

yj-soh commented on src/main/java/teammates/common/datatransfer/InstructorAttributes.java:
```diff
@@ -311,4 +311,20 @@ public boolean isEqualToAnotherInstructor(InstructorAttributes instructor) {
     }
+    
+    public boolean isCustomRole() {
+        return role.equals(Const.InstructorPermissionRoleNames.INSTRUCTOR_PERMISSION_ROLE_CUSTOM);
```
Will `role` ever be `null`?
Reversing it may be more safe.
Calling `equals()` on a null `role` will cause a NPE, but reversing them will make it only return `false`.

yj-soh commented on src/main/java/teammates/ui/controller/InstructorCourseEditPageData.java:
```diff
@@ -7,24 +8,129 @@
+        this.currentInstructor = currentInstructor;
+        
+        createButtons();
+        createInstructorPanelList(instructorList, sectionNames, feedbackNames);
```
It should be best to connect `createInstructorPanelList()` with its object initialisation.
Try either moving the `instructorPanelList = new ArrayList<...` into the method, or do a `instructorPanelList = createInstructorPanelList(...` here like what was done for `addInstructorPanel` below.

yj-soh commented on src/main/java/teammates/ui/template/CourseEditInstructorPanel.java:
```diff
@@ -0,0 +1,210 @@
+        createSectionRows(instructorIndex, sectionNames, feedbackNames);
+        permissionInputGroup1 = createPermissionInputGroup1ForInstructorPanel(instructor);
+        permissionInputGroup2 = createPermissionInputGroup2ForInstructorPanel(instructor);
+        permissionInputGroup3 = createPermissionInputGroup3ForInstructorPanel(instructor);
```
Keep it consistent whether to create the class-scope object in the method, or to create a method-scope object and pass it back.

yj-soh commented on src/main/java/teammates/ui/template/CourseEditInstructorPanel.java:
```diff
@@ -0,0 +1,210 @@
+    }
+    
+    private ElementTag createCheckBox(String content, InstructorAttributes instructor, String privilege) {
+        boolean isChecked = (instructor == null) || instructor.isAllowedForPrivilege(privilege);
```
The class-scope `instructor` can be used here instead.
Just wondering, though, but is it possible for instructor to be `null` here?

bamboo3250 replied:
Yes, it can. It happens for the addInstructorPanel, where there is no instructor and the panel still looks like those with instructor.

yj-soh commented on src/main/webapp/WEB-INF/tags/instructor/course/courseEditAddInstructorPanel.tag:
```diff
@@ -0,0 +1,285 @@
+        <form method="post" action="<%=Const.ActionURIs.INSTRUCTOR_COURSE_INSTRUCTOR_ADD%>" name="formAddInstructor" 
+                class="form form-horizontal" id="formAddInstructor">
+            <input type="hidden" name="<%=Const.ParamsNames.COURSE_ID%>" value="${courseId}">
+            <input type="hidden" name="<%=Const.ParamsNames.USER_ID%>" value="${googleId}">
```
If this is `data.account.googleId`, there's no need to pass it in.

yj-soh commented on src/main/webapp/WEB-INF/tags/instructor/course/courseEditAddInstructorPanel.tag:
```diff
@@ -0,0 +1,285 @@
+                    <div class="col-sm-9"><input class="form-control" type="text"
+                        name="<%=Const.ParamsNames.INSTRUCTOR_EMAIL%>" id="<%=Const.ParamsNames.INSTRUCTOR_EMAIL%>"
+                        data-toggle="tooltip" data-placement="top" title="Enter the Email of the instructor."
+                        maxlength=<%=FieldValidator.EMAIL_MAX_LENGTH%> tabindex=9/>
```
Should these be enclosed by quotation marks?

bamboo3250 replied:
Yes, that's what I mentioned yesterday. Will fix these.

yj-soh commented on src/main/java/teammates/ui/template/CourseEditInstructorPanel.java:
```diff
@@ -0,0 +1,210 @@
+    private List<CourseEditSectionRow> sectionRows;
+    
+    public CourseEditInstructorPanel(int instructorToShowIndex, int instructorIndex, 
+                                     InstructorAttributes instructor, InstructorAttributes currentInstructor,
```
Is `currentInstructor` used?

bamboo3250 commented:
@yj-soh fixed. Please review again.

yj-soh commented on src/main/java/teammates/ui/template/CourseEditInstructorPanel.java:
```diff
@@ -0,0 +1,210 @@
+        createSectionRows(instructorIndex, sectionNames, feedbackNames);
+        permissionInputGroup1 = createPermissionInputGroup1ForInstructorPanel();
+        permissionInputGroup2 = createPermissionInputGroup2ForInstructorPanel();
+        permissionInputGroup3 = createPermissionInputGroup3ForInstructorPanel();
```
Maybe make `createSectionRows` return the variable for consistency?

yj-soh commented on src/main/java/teammates/common/datatransfer/InstructorAttributes.java:
```diff
@@ -311,4 +311,20 @@ public boolean isEqualToAnotherInstructor(InstructorAttributes instructor) {
     }
+    
+    public boolean isCustomRole() {
+        return (role != null) && role.equals(Const.InstructorPermissionRoleNames.INSTRUCTOR_PERMISSION_ROLE_CUSTOM);
```
`Const.InstructorPermissionRoleNames.INSTRUCTOR_PERMISSION_ROLE_CUSTOM.equals(role)`
should give the same result.

bamboo3250 replied:
Oh, i get it. Never thought using a constant like that :+1:

bamboo3250 commented:
@yj-soh fixed

yj-soh commented on src/main/webapp/jsp/instructorCourseEdit.jsp:
```diff
@@ -1,973 +1,36 @@
-                                                <div class="col-sm-3">
-                                                    <input type="checkbox" name="<%=Const.ParamsNames.INSTRUCTOR_PERMISSION_MODIFY_STUDENT%>"
-                                                            value="true" checked="checked" /> Enroll/Edit/Delete Students
-                
```
Is this still needed?
Or can it be directly accessed from that file?

yj-soh commented:
Should be fine after that's done.

bamboo3250 commented:
@yj-soh fixed.

yj-soh commented on src/main/java/teammates/ui/template/CourseEditSectionRow.java:
```diff
@@ -0,0 +1,196 @@
+        for (String feedbackName : feedbackNames) {
+            String[] privileges = {Const.ParamsNames.INSTRUCTOR_PERMISSION_SUBMIT_SESSION_IN_SECTIONS,
+                                   Const.ParamsNames.INSTRUCTOR_PERMISSION_VIEW_SESSION_IN_SECTIONS,
+                                   Const.ParamsNames.INSTRUCTOR_PERMISSION_MODIFY_SESSION_COMMENT_IN_SECTIONS};
```
This should be declared outside the `for` loop; no point recreating it each time.

yj-soh commented on src/main/java/teammates/ui/template/CourseEditSectionRow.java:
```diff
@@ -0,0 +1,196 @@
+                  boolean isChecked = (instructor != null) && instructor.isAllowedForPrivilege(sectionName, 
+                                                                                               feedbackName,
+                                                                                               privilege);
+                  checkBoxList.add(createCheckBox(null, name, "true", isChecked));
```
Your indentation is a little broken here.

yj-soh commented on src/test/java/teammates/test/cases/ui/InstructorCourseEditPageActionTest.java:
```diff
@@ -128,7 +130,16 @@ public void testExecuteAndPostProcess() throws Exception {
-
+    
+    public List<InstructorAttributes> getInstructorList(List<CourseEditInstructorPanel> instructorPanelList) {
+        ArrayList<InstructorAttributes> instructorList = new ArrayList<InstructorAttributes>();
```
`List<InstructorAttributes>`.

As an alternative, you may find it tidier to simply modify `verifySameInstructorList()` to support a `List<CourseEditInstructorPanel>` passed in instead of making a method to extract a list of instructors.

yj-soh commented on src/test/java/teammates/test/cases/ui/pagedata/InstructorCourseEditPageDataTest.java:
```diff
@@ -0,0 +1,80 @@
+        assertEquals(4, pageData.getInstructorPanelList().get(0).getPermissionInputGroup2().size());
+        assertEquals(3, pageData.getInstructorPanelList().get(0).getPermissionInputGroup3().size());
+        assertEquals(sectionNames.size(), pageData.getInstructorPanelList().get(0).getSectionRows().size());
+        assertEquals(feedbackSessionNames.size(), pageData.getInstructorPanelList().get(0).getSectionRows().get(0).getFeedbackSessions().size());
```
Maybe store this in a variable instead of calling `pageData.getInstructorPanelList().get(0)` every time?

yj-soh commented on src/test/java/teammates/test/cases/ui/pagedata/InstructorCourseEditPageDataTest.java:
```diff
@@ -0,0 +1,80 @@
+        assertEquals(3, pageData.getAddInstructorPanel().getPermissionInputGroup3().size());
+        assertEquals(sectionNames.size(), pageData.getAddInstructorPanel().getSectionRows().size());
+        assertEquals(feedbackSessionNames.size(), pageData.getAddInstructorPanel().getSectionRows().get(0).getFeedbackSessions().size());
+        
```
Is this sufficient?
It seems to miss the instructor remind button branch entirely.

bamboo3250 commented:
@yj-soh I added more test cases and added a new instructor into typicalDataBundle.json so I need to change some other tests as well. Please review it.

yj-soh commented on src/test/java/teammates/test/cases/ui/pagedata/InstructorCourseEditPageDataTest.java:
```diff
@@ -0,0 +1,146 @@
+        assertEquals(feedbackSessionNames.size(), sectionRow.getFeedbackSessions().size());
+        assertNotNull(sectionRow.getToggleSessionLevelInSectionButton());
+        assertFalse(sectionRow.isSectionSpecial());
+        assertEquals((sectionNames.size() - 1) / 3 + 1, sectionRow.getSpecialSections().size());
```
Where does `(sectionNames.size() - 1) / 3 + 1` come from?
Maybe a comment or variable to explain?

bamboo3250 replied:
Those sections are separated by a group of 3 so that is the formula for the number of groups. I will add a comment for that.

kanghj commented on src/test/resources/data/typicalDataBundle.json:
```diff
@@ -269,6 +269,17 @@
+        },
+        "instructorNotYetJoinCourse1":{
+            "googleId":null,
+            "courseId":"idOfTypicalCourse1",
```
Can you move this nearer to the top, where the other instructors for the course idOfTypicalCourse1 are? The rest of the instructors are not very neatly organised, but I guess we shouldn't mess things up more. 

Have you tried running the other tests? Many tests use the typicalDataBundle and modifications to this file may affect the other tests.

bamboo3250 replied:
Yes, I ran all tests so I made modifications to all tests that were affected.

kanghj replied:
Ok, great! This should be fine then.

bamboo3250 commented:
@yj-soh please check again

yj-soh commented on src/main/java/teammates/ui/controller/InstructorCourseEditPageData.java:
```diff
@@ -7,24 +8,130 @@
+    private CourseAttributes course;
+    private List<CourseEditInstructorPanel> instructorPanelList;
+    private CourseEditInstructorPanel addInstructorPanel;
+    private InstructorAttributes currentInstructor;
```
If you're only using `currentInstructor` in the method scopes, maybe pass it around the methods instead of storing it in the class scope?

yj-soh commented on src/main/webapp/WEB-INF/tags/instructor/course/courseEditInstructorListPanelBody.tag:
```diff
@@ -0,0 +1,98 @@
+                    <input class="form-control" type="text" name="<%=Const.ParamsNames.INSTRUCTOR_NAME%>" 
+                            id="<%=Const.ParamsNames.INSTRUCTOR_NAME%>${instructorPanel.index}" value="${instructorPanel.instructor.name}" 
+                            data-toggle="tooltip" data-placement="top" title="Enter the name of the instructor." 
+                            maxlength=<%=FieldValidator.PERSON_NAME_MAX_LENGTH%> tabindex="4" disabled="disabled">
```
You may want to fix such occurrences where the value of certain attributes aren't surrounded in quotations marks in this PR as well.
(`maxlength=<%=FieldValidator.PERSON_NAME_MAX_LENGTH%>`)

yj-soh commented:
Looks okay, feel free to merge after these comments.
Be sure to test.

bamboo3250 commented:
okay, thanks
